src/backend/distributed/commands/alter_table.c

```diff
@@ -0,0 +1,1370 @@
+}
+
+
+/*
+ * ConvertTable is used for converting a table into a new table with different properties.
+ * The conversion is done by creating a new table, moving everything to the new table and
+ * dropping the old one. So the oid of the table is not preserved.
+ *
+ * The new table will have the same name, columns and rows. It will also have partitions,
+ * views, sequences of the old table. Finally it will have everything created by
+ * GetPostLoadTableCreationCommands function, which include indexes. These will be
+ * re-created during conversion, so their oids are not preserved either (except for
+ * sequences). However, their names are preserved.
+ *
+ * The dropping of old table is done with CASCADE. Anything not mentioned here will
+ * be dropped.
+ *
+ * The function returns a TableConversionReturn object that can stores variables that
+ * can be used at the caller operations.
+ */
+TableConversionReturn *
+ConvertTable(TableConversionState *con)
+{
+	if (con->conversionType == UNDISTRIBUTE_TABLE && con->cascadeViaForeignKeys)
+	{
+		CascadeOperationForConnectedRelations(con->relationId, ExclusiveLock, CASCADE_FKEY_UNDISTRIBUTE_TABLE);
+	
+		/*
+		 * Undistributed every foreign key connected relation in our foreign key
+		 * subgraph including itself, so return here.
+		 */
+		return NULL;
+	}
+	char *newAccessMethod = con->accessMethod ? con->accessMethod :
+							con->originalAccessMethod;
+	List *preLoadCommands = GetPreLoadTableCreationCommands(con->relationId, true,
+															newAccessMethod);
+
+	bool includeIndexes = true;
+	if (con->accessMethod && strcmp(con->accessMethod, "columnar") == 0)
+	{
+		ereport(NOTICE, (errmsg("any index will be dropped, because "
+								"columnar tables cannot have indexes")));
+		includeIndexes = false;
+	}
+	List *postLoadCommands = GetPostLoadTableCreationCommands(con->relationId,
+															  includeIndexes);
+	List *justBeforeDropCommands = NIL;
+	List *attachPartitionCommands = NIL;
+
+	postLoadCommands = list_concat(postLoadCommands,
+								   GetViewCreationCommandsOfTable(con->relationId));
+
+	List *foreignKeyCommands = NIL;
+	if (con->conversionType == ALTER_DISTRIBUTED_TABLE)
+	{
+		foreignKeyCommands = GetForeignConstraintToReferenceTablesCommands(
+			con->relationId);
+		if (con->cascadeToColocated == CASCADE_TO_COLOCATED_YES ||
+			con->cascadeToColocated == CASCADE_TO_COLOCATED_NO_ALREADY_CASCADED)
+		{
+			List *foreignKeyToDistributedTableCommands =
+				GetForeignConstraintToDistributedTablesCommands(con->relationId);
+			foreignKeyCommands = list_concat(foreignKeyCommands,
+											 foreignKeyToDistributedTableCommands);
+
+			List *foreignKeyFromDistributedTableCommands =
+				GetForeignConstraintFromDistributedTablesCommands(con->relationId);
+			foreignKeyCommands = list_concat(foreignKeyCommands,
+											 foreignKeyFromDistributedTableCommands);
+		}
+		else
+		{
+			WarningsForDroppingForeignKeysWithDistributedTables(con->relationId);
+		}
+	}
```

Comment:
in the else if (con->conversionType == ALTER_TABLE_SET_ACCESS_METHOD) case I could have something like a GetForeignConstraintToOtherTablesCommands that includes all table tables.